src/Cryptol/Eval/Generic.hs

```diff
@@ -319,89 +324,229 @@ arithNullary sym opw opi opz = loop
+fromIntegerV sym =
+  tlam $ \ a ->
+  lam  $ \ v ->
+  do i <- fromVInteger <$> v
+     intV sym i a
 
 {-# INLINE addV #-}
 addV :: Backend sym => sym -> Binary sym
-addV sym = arithBinary sym opw opi opz
+addV sym = ringBinary sym opw opi opz opq
   where
     opw _w x y = wordPlus sym x y
     opi x y = intPlus sym x y
     opz m x y = znPlus sym m x y
+    opq x y = rationalAdd sym x y
 
 {-# INLINE subV #-}
 subV :: Backend sym => sym -> Binary sym
-subV sym = arithBinary sym opw opi opz
+subV sym = ringBinary sym opw opi opz opq
   where
     opw _w x y = wordMinus sym x y
     opi x y = intMinus sym x y
     opz m x y = znMinus sym m x y
+    opq x y = rationalSub sym x y
+
+{-# INLINE negateV #-}
+negateV :: Backend sym => sym -> Unary sym
+negateV sym = ringUnary sym opw opi opz opq
+  where
+    opw _w x = wordNegate sym x
+    opi x = intNegate sym x
+    opz m x = znNegate sym m x
+    opq x = rationalNegate sym x
 
 {-# INLINE mulV #-}
 mulV :: Backend sym => sym -> Binary sym
-mulV sym = arithBinary sym opw opi opz
+mulV sym = ringBinary sym opw opi opz opq
   where
     opw _w x y = wordMult sym x y
     opi x y = intMult sym x y
     opz m x y = znMult sym m x y
+    opq x y = rationalMul sym x y
+
+--------------------------------------------------
+-- Integral
 
 {-# INLINE divV #-}
 divV :: Backend sym => sym -> Binary sym
-divV sym = arithBinary sym opw opi opz
+divV sym = integralBinary sym opw opi
   where
     opw _w x y = wordDiv sym x y
     opi x y = intDiv sym x y
-    opz m x y = znDiv sym m x y
+
+{-# SPECIALIZE expV :: Concrete -> GenValue Concrete #-}
+expV :: Backend sym => sym -> GenValue sym
+expV sym =
+  tlam $ \aty ->
+  tlam $ \ety ->
+   lam $ \am -> return $
+   lam $ \em ->
+     do a <- am
+        e <- em
+        case ety of
+          TVInteger ->
+            let ei = fromVInteger e in
+            case integerAsLit sym ei of
+              Just n
+                | n == 0 ->
+                   do onei <- integerLit sym 1
+                      intV sym onei aty
+
+                | n > 0 ->
+                    do ebits <- enumerateIntBits' sym n ei
+                       computeExponent sym aty a ebits
+
+                | otherwise -> raiseError sym NegativeExponent
+
+              Nothing -> liftIO (X.throw (UnsupportedSymbolicOp "integer exponentiation"))
+
+          TVSeq _w el | isTBit el ->
+            do ebits <- enumerateWordValue sym =<< fromWordVal "(^^)" e
+               computeExponent sym aty a ebits
+
+          _ -> evalPanic "expV" [show ety ++ " not int class `Integral`"]
+
+
+{-# SPECIALIZE computeExponent ::
+      Concrete -> TValue -> GenValue Concrete -> [SBit Concrete] -> SEval Concrete (GenValue Concrete)
+  #-}
+computeExponent :: Backend sym =>
+  sym -> TValue -> GenValue sym -> [SBit sym] -> SEval sym (GenValue sym)
+computeExponent sym aty a bs0 =
+  do onei <- integerLit sym 1
+     one <- intV sym onei aty
+     loop one bs0
```

Comment:
Note to self: maybe it makes sense to drop concrete leading zeros here before going into the loop?